### ISSUE TITLE
Build broken with vvp master

### ISSUE THREAD

**Issue by lubosz** [nyorain/vvg#1]
Hi, interesting library. Which commit of vpp do I need to build vvg?
Tried some stuff but can't figure it out.

You probably should tag vvg compatiblity in vvp and reference that tag in the submodule.

Cheers.

**Comment by nyorain** [comment]
hey, thank you! I did not work on this for quite a while. 
Will update and test it for the latest vpp version this weekend.
And yeah, then it would probably be a good idea to use a fixed vpp tag/commit.

There are still a few features missing for a full nanovg implementation though (mostly related to stencil buffer usage).

**Comment by ec1oud** [comment]
Well now it seems you have converted all your other projects from cmake to meson... but not this one.  Regardless whether I try to revert the others to some time around Feb. 18 (when they all still had cmake) or convert this one to meson, it's a can of worms.  My fork has my attempt to convert to meson; maybe it's a starting point?  I don't understand why meson thinks vpp has a dependency called "vpp" and then complains it can't find it.

```
[0/1] Regenerating build files.
The Meson build system
Version: 0.43.0
Source dir: /home/rutledge/src/vulkan/vvg
Build dir: /home/rutledge/src/vulkan/vvg/build
Build type: native build
Project name: vvg
Native C compiler: ccache cc (gcc 7.2.0)
Native C++ compiler: ccache c++ (gcc 7.2.0)
Build machine cpu family: x86_64
Build machine cpu: x86_64
Found pkg-config: /usr/bin/pkg-config (0.29.2)

Executing subproject vpp.

Project name: vpp
Native C compiler: ccache cc (gcc 7.2.0)
Native C++ compiler: ccache c++ (gcc 7.2.0)
Dependency threads found: YES
Also couldn't find a fallback subproject in subprojects/dlg for the dependency dlg
Also couldn't find a fallback subproject in subprojects/vpp for the dependency vpp

Meson encountered an error in file meson.build, line 22, column 0:
Native dependency 'vpp' not found

A full log can be found at /home/rutledge/src/vulkan/vvg/build/meson-logs/meson-log.txt
FAILED: build.ninja 
/usr/bin/python /usr/bin/meson --internal regenerate /home/rutledge/src/vulkan/vvg /home/rutledge/src/vulkan/vvg/build --backend ninja
ninja: error: rebuilding 'build.ninja': subcommand failed
```

**Comment by nyorain** [comment]
Hey, the problem is in the dependency handling of meson. It will not handle subprojects recursively (this was done intentionally, head over to the meson guys for more information) so this library would have wrap files of vpp in its subprojects folder.

I never got to really fix this project, I'm sorry. I will push the update i did a few months ago (i got it working back then with meson) but since i reworked vpp, it won't work now.
This whole library was hacked together in a few days, it lacks support for many things so fixing everything would take some time and I'm not so sure anymore this is even a good idea (we cannot really use the main advantages of vulkan with the rather opengl-focused nanovg interface). Will update the README to reflect the current state, sorry if you wasted any time on this.

**Comment by ec1oud** [comment]
I can believe that there's some impedance mismatch, but what do you think is the right way to do that kind of 2D stuff with Vulkan then?  I've just been casting about for some inspiration, and pursuing the vague idea that the CPU/GPU boundary ought to correspond better to the model-view boundary: don't imperatively render every frame with a fresh set of vertices, but come up with a way to upload as much as possible of the "view" ahead of time and then update it incrementally later.  But I guess that's going a different direction than what the game engines have been doing.

**Comment by nyorain** [comment]
I also made some thoughts about this in the past. And yeah it boils down to not record a commandBuffer every frame (which we have to do with the classical nanovg interface) and instead try to do everything without rerecording/recalculating/reuploading (you can usually update images, vertex- and uniform buffers, i.e. the path and style of a shape without rerecording).

The best way would probably an OOP api, where you have objects of classes like Path, Text or Brush that you can update independently and only rerecord the command buffer if you need to (e.g. if a buffer had to be recreated or you added a new object to render). This way if nothing changes (no animations and stuff) you would use almost no cpu in the rendering logic.

Basically transform the nanovg api into an OOP api with objects that keep their state over multiple frames. I would like to transform this library into something like that but it will take some time.

**Comment by nyorain** [comment]
Just in case anyone stumbles about this and is still interested:
I followed that approach and built a high level retained mode graphics api for vulkan: [nyorain/rvg](https://github.com/nyorain/rvg)
